src/libos/crates/async-rt/src/wait/waiter.rs

```diff
@@ -155,3 +170,67 @@ impl<'a> Future for WaitFuture<'a> {
         }
     }
 }
+
+impl<'a> Drop for WaitFuture<'a> {
+    fn drop(&mut self) {
+        let mut raw_waker = self.waiter.raw_waker.lock();
+        if let WaiterState::Waiting = self.waiter.state() {
+            *raw_waker = None;
+            self.waiter.set_state(WaiterState::Idle);
+        }
+    }
+}
+
+/// A wrapper to unify the future type.
+/// This wrapper can support two future types and store only one future.
+#[pin_project]
+pub struct WrapperFuture<F1, F2> {
+    #[pin]
+    future: UnionFuture<F1, F2>,
+}
+
+/// A pinned enum to support union format.
+/// Since it's constructed by the pin_project crate, the real name of this enum is `UnionFutureProj`.
+#[pin_project(project = UnionFutureProj)]
+pub enum UnionFuture<F1, F2> {
```

Comment:
The doc of `select!`: Polls multiple futures and streams simultaneously.
What we want is a union future type, which is only one future, instead of multiple futures.
ps, I got the idea of the `UnionFuture` enum from the async_std crate.

Reply:
Use `select_biased` macro
``` Rust
select_biased! {
                    _ = self.inner.wait().fuse() => { ... }
                    _ = TimerFutureEntry::new(&timer_entry).fuse() => { ... }
};
```